[BOLT][DWARF][NFC] Split processUnitDIE into two lambdas

sayhaan opened the pull request:
Split processUnitDIE into two lambdas to separate the processing of DWO CUs and CUs in the main binary.

sayhaan committed:
Split processUnitDIE into two lambdas

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D59822423

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -768,17 +773,35 @@ void DWARFRewriter::updateDebugInfo() {
+      }
+      for (DWARFUnit *CU : CompileUnits) {
+        std::optional<uint64_t> RangesBase;
+        if (RangesBaseByCU.count(CU->getOffset()))
```
User iterator to check and get if exists.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
+          auto AddrW =
+              std::make_unique<DebugAddrWriter>(&BC, CU.getAddressByteSize());
+          AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+          LocListWritersByCU[CUIndex] = std::make_unique<DebugLocWriter>();
```
Maybe it's best to simplify this and just created instance of Loc/loclist writer when we iterate for Main Binary and store it in vector.

sayhaan replied:
I'm not sure if we can do that because the instance is created in createRangeLocListAddressWriters which also populates the Address/Ranges writers maps and uses the Address Writer, which processSplitCU uses. If move it to the loop where we iterate for the main binary then processSplitCU won't work.

ayermolo replied:
Ah right, dwarf5 instance takes in.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
+            : std::optional<std::string>(opts::DwarfOutputPath.c_str());
+    {
+      std::lock_guard<std::mutex> Lock(AccessMutex);
+      DWOName = DIEBlder->updateDWONameCompDir(
```
This common state that will require lock. In followup diff move it outside of this lambda.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
+      RangesBase = Unit->getVersion() >= 5
+                       ? RangeListsSectionWriter.get()->getSectionOffset()
+                       : LegacyRangesSectionWriter.get()->getSectionOffset();
+      setDwoRangesBase(*DWOId, *RangesBase);
```
Hmm actually this whole setDwoRangesBase etc no longer used anywhere....
Can you remove it in a different PR. Rebase this one on top of it.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
+      TempRangesSectionWriter = RangeListsWritersByCU[*DWOId].get();
+    } else {
+      TempRangesSectionWriter = LegacyRangesWritersByCU[*DWOId].get();
+      RangesBase = Unit->getVersion() >= 5
```
I don't think RangesBase needs to be set here. It's only needed when we process CUs in main binary, so we know when to update it, or create a new one.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -767,17 +769,28 @@ void DWARFRewriter::updateDebugInfo() {
     for (std::unique_ptr<DWARFUnit> &CU : BC.DwCtx->compile_units())
-      ThreadPool.async(processUnitDIE, CU.get(), &DIEBlder);
-    ThreadPool.wait();
+      // ThreadPool.async(processUnitDIE, CU.get(), &DIEBlder);
```
Remove this change

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -767,17 +769,28 @@ void DWARFRewriter::updateDebugInfo() {
       DIEBlder.buildCompileUnits(Vec);
-      for (DWARFUnit *CU : DIEBlder.getProcessedCUs())
-        processUnitDIE(CU, &DIEBlder);
+      llvm::SmallVector<DWARFUnit *> CompileUnits(
```
Why is this necessary?

sayhaan replied:
Since we have two loops we only call DIEBlder.getProcessedCUs() once instead of in both loops.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,119 +632,121 @@ void DWARFRewriter::updateDebugInfo() {
+                   DebugLocDWoWriter, DWOStrOffstsWriter, DWOStrWriter,
+                   GDBIndexSection);
+  };
+  auto processMainBinaryCU = [&](DWARFUnit *Unit, DIEBuilder *DIEBlder,
```
Since we are modifying this code anyway. In this and other lambda change to pass by reference.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -632,119 +632,121 @@ void DWARFRewriter::updateDebugInfo() {
+                std::move(DWORangeListsSectionWriter);
+          }
+          AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+
```
remove space

sayhaan committed:
[BOLT][DWARF] Remove deprecated opt

Summary: Remove deprecated DeterministicDebugInfo option and its uses.

Test Plan: Run `ninja llvm check-bolt.`

Reviewers: ayermolo, aaupov, #llvm-bolt

Reviewed By: ayermolo

Differential Revision: https://phabricator.intern.facebook.com/D59942763

Tasks: T187681160

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
+            ? std::nullopt
+            : std::optional<std::string>(opts::DwarfOutputPath.c_str());
+    {
+      std::lock_guard<std::mutex> Lock(AccessMutex);
```
Remove in followup PR.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
-                     GDBIndexSection);
+    if (!SplitCU)
+      return;
+    DebugAddrWriter &AddressWriter =
```
Move the checks outside, and just pass in AddressWriter and SplitCU.
My thinking is less reliance on shared data structures the better.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
+        DWOStrOffstsWriter, DWOStrWriter, **SplitCU, DwarfOutputPath, DWOName);
+    DebugLoclistWriter DebugLocDWoWriter(Unit, Unit.getVersion(), true,
+                                         AddressWriter);
+    DebugRangesSectionWriter *TempRangesSectionWriter =
```
Also this, move check outside of lambda and pass in.

ayermolo commented on bolt/lib/Rewrite/DWARFRewriter.cpp:
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
-      }
-      AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+      [&](DWARFUnit &CU,
+          llvm::DenseMap<uint64_t, uint64_t> &LocListWritersIndexByCU) {
```
I think you can just capture it by reference in lambda.

sayhaan replied:
For LocListWritersIndexByCU? Wouldn't it not be in the same scope since I initialize it inside the for loop? I could move it outside but that would increase the scope it covers when it's only used in one lambda.

sayhaan replied:
For LocListWritersIndexByCU? It wouldn't be in scope for the lambda to capture by reference since I initialize it inside the for loop. I could move it outside but then its lifetime is longer than it needs to be since it's only used in the lambda.

sayhaan committed:
Formatting changes

sayhaan committed:
Merge branch 'main' into T187681160-1

aaupov commented:
Heads up that this diff is not NFC: https://lab.llvm.org/buildbot/#/builders/92/builds/2042
